### ISSUE TITLE
check use of "origin_long_name"

### ISSUE THREAD

**Issue by thareUSGS** [USGS-Astrogeology/GDAL_scripts#34]
within: https://github.com/USGS-Astrogeology/GDAL_scripts/blob/master/OGC_IAU2000_WKT_v2/Source_Python/create_IAU2000.py

The new addition of "origin_long_name" field (from table below) might be getting used incorrectly. This is typically not the "Reference_Meridian" as understood by the a WKT projection. These names represent the "Wo" value which is not always defined as 0. For example, on Mercury, Hun Kal is really 20 degrees from 0.0. Now many are defined as zero, but where it is not, I'm not sure we can use it for Reference_Meridian. See: https://github.com/USGS-Astrogeology/GDAL_scripts/blob/master/OGC_IAU2000_WKT_v2/Docs/IAU2015_Archinaletal2018.pdf

ideas:
(1) Go back to simply using "Reference_Meridian", 0.0 for all WKT definitions
(2) Add an additional field (e.g. REMARK) which defines the offset to that feature (from the publication). And then in the code, if indeed it is 0.0, then the name can be used. If not zero, then fall back to "Reference_Meridian"

table: https://github.com/USGS-Astrogeology/GDAL_scripts/blob/master/OGC_IAU2000_WKT_v2/naifcodes_radii_m_wAsteroids_IAU2015.csv

**Comment by thareUSGS** [comment]
I am really afraid for the implications of setting a non-zero Reference Meridian as I believe it will greatly impact the interoperability of existing (and released) data (e.g. using PRIMEM["Hun Kal", 20] ).

This is not something defined or available in the original PDS labels. But by defining Hun Kal to 20 degrees in the PDS product, we now require mapping applications to support a messy Geographic Transformation ( http://wiki.gis.com/wiki/index.php/Geographic_Transformation ). This would meaning downloading all global map projected PDS products and then shifting the pixels in a global mosaic by 20 degrees (or possible play other games) to set PRIMEM["Hun Kal", 20] .

For better or worse, I really try to take interoperability into account (even at the expense of being 100% accurate for say a definition).  By setting Mercury to a ["Reference Meridian", 0.0], we are still syntactically accurate although we have obfuscated the use of Hun Kal. 

Maybe I have the impact all wrong, perhaps Martin can chime in.

**Comment by desruisseaux** [comment]
I would expect `PRIMEM["Hun Kal", 20]` to work with the GDAL version under development. It should work at least with Apache SIS, and the new Proj/GDAL library reuses many design principles from SIS. It does not necessarily implies a geographic transformation between Coordinate Reference Systems (CRS). I do not know what look like previous definitions, but I would expect the following consequences:

* If the `DATUM` element in old and new CRS definitions have the same name, the map projection library may possibly consider that there is no geographic transformations. Apache SIS and - I think - new GDAL version use the `DATUM` name for determining if there is a datum shift or not. To be strict, if the `PRIMEM` inside the `DATUM` element is different, it should be a different datum. But if we consider that change of value in `PRIMEM` as a correction of an existing `DATUM` instead than a new datum, maybe it will work.
* A change of `PRIMEM` value in a projected CRS may not necessarily require to reproject the data. It depends if the `central_meridian` projection parameter compensate for the prime meridian change.
* In the worst case, if there is a decision to really use distinct `DATUM` and CRS definitions, it is possible to define the new CRS using `PRIMEM["Hun Kal", 20]` and explicitly defines its relationship to the old CRS. That relationship can be defined as no-operation. This mechanism is provided in WKT 2 by the `BOUNDCRS` element.

Note that `TOWGS84` element is deprecated and does not exist anymore in WKT 2. It is replaced by the above-cited `BOUNDCRS` element. The legacy `TOWGS84` element was a way to define the relationship between a CRS and WGS84. The new `BOUNDCRS` element is a generalization of this mechanism, which allow to define the relationship to an arbitrary CRS instead than only WGS84.

**Comment by thareUSGS** [comment]
I understand the ideas above should work, but in current software it just doesn't or is not reliable. I just think we are opening up too many potential issues. Perhaps supporting these ideas in WKT2 might be okay but I do not think I can support them for WKT(v1) and current software.  

To your points:

- _"have the same name, the map projection library may possibly consider that there is no geographic transformations"_

Even when the reference meridian have different names, the software I tested does not care. It will treat them as the same (or ignore it completely). See the impact in QGIS (below). The Hun Kal defined reference meridian is shifted 20 degrees from where it should be and it takes much longer to draw. Not only is the map correctly shifted, but the edges get clipped.  

Also if the software translates the WKT (or coded definition) to proj4, then it won't have the string available. Here is the original and modified test case.
```
modified: +proj=eqc +lat_ts=0 +lat_0=0 +lon_0=180 +x_0=0 +y_0=0 +a=2439700 +b=2439700 **+pm=20** +units=m +no_defs
original  : +proj=eqc +lat_ts=0 +lat_0=0 +lon_0=180 +x_0=0 +y_0=0 +a=2440000 +b=2440000 +units=m +no_defss
```

- _"depends if the central_meridian projection parameter compensate for the prime meridian change"_

Could we play games with redefining the central meridian, Yes, but I think that will just exacerbate the problem and be less interoperable.

- _"This mechanism is provided in WKT 2 by the BOUNDCRS element."_

So perhaps the answer is to try to implement this in WKT2.  But I still do not think I can support this use case in WKT (v1) definitions.

See this Gif, the only change in the map projection is to define Hun Kal as 20.0, which shows QGIS incorrectly displaying the image. 

![mercury_hunkal20](https://user-images.githubusercontent.com/3095138/52166423-3d95fd80-26ca-11e9-82fb-eeead0a94668.gif)

**Comment by J-Christophe** [comment]
Ok, we set the reference meridian to zero. Il will do the change in the
python script.
In addition, i will write code for wkt2. So for wkt2, should i put "Hun
Kal" in the remark field ?




Le sam. 2 févr. 2019 à 17:19, Trent Hare <notifications@github.com> a
écrit :

**Comment by desruisseaux** [comment]
On the other hand, providing information as accurate as possible can reduce problems in medium or long terms. Especially since the observed problems concern one particular software, not necessarily GIS software in general. QGIS is based on GDAL/Proj.4, which has a flawed support of referencing systems and WKT in versions before the "gdalbarn" effort. [The community has raided 144,000$ for addressing those problems](https://gdalbarn.com/) and the progress are good. The next GDAL version will be much better than the current one regarding coordinate operations and WKT. If there is any issue in interpreting `PRIMEM["Hun Kal", 20]` with new GDAL and if this issue is reported before the end of "gdalbarn" effort, I'm pretty much convinced that Even Rouault would be ready to fix it.

My proposal: the reference meridian could be set to zero for the WKT 1 format (I'm neutral on that), for compatibility with current QGIS version. But for the WKT 2 format, I suggest to "do the right thing". When we have a WKT 2 ready for Mercury with `PRIMEM["Hun Kal", 20]` , we could ask to Even Rouault if the new GDAL is okay with that WKT 2. If Even confirms that GDAL is okay, QGIS, MapServer and PostGIS should follow (after they have been upgraded).

**Comment by thareUSGS** [comment]
100% agree. For WKT1 let's keep it working using 0.0. But for WKT2, lets get the PRIMEM in there and see how it works.

**Comment by rouault** [comment]
@J-Christophe pinged me on this but I've some difficulties understand what the exact concern is here without concrete/reproducable scenarios, and what should potentialy be fixed.
I encourage folks to build PROJ master from https://github.com/OSGeo/proj.4/ and report there any issues they might find.
I assume GDAL WKT1 should already properly honour the PrimeMeridian concept (perhaps except if the angular unit != degree) and translate it to the +pm proj keyword. WKT2 support doesn't change much in that area. The longitude parameters of the projection should be relative to the PrimeMeridian.

Apparently there's some discussion about having DATUM objects having the same name but different PrimeMeridian values. Before testing, I wasn't sure to remember what the new PROJ implementation would do, but apparently it doesn't just take into account the name to determine Datum equality.

So if you have src.wkt with PRIMEM["Hun Kal",20]
```
PROJCRS["A",
    BASEGEOGCRS["Mercury",
        DATUM["Mercury",
            ELLIPSOID["Mercury",2439700,0,
                LENGTHUNIT["metre",1,
                    ID["EPSG",9001]]]],
        PRIMEM["Hun Kal",20,
            ANGLEUNIT["degree",0.0174532925199433,
                ID["EPSG",9122]]]],
    CONVERSION["unknown",
        METHOD["Equidistant Cylindrical (Spherical)",
            ID["EPSG",1029]],
        PARAMETER["Latitude of 1st standard parallel",0,
            ANGLEUNIT["degree",0.0174532925199433],
            ID["EPSG",8823]],
        PARAMETER["Longitude of natural origin",180,
            ANGLEUNIT["degree",0.0174532925199433],
            ID["EPSG",8802]],
        PARAMETER["False easting",0,
            LENGTHUNIT["metre",1],
            ID["EPSG",8806]],
        PARAMETER["False northing",0,
            LENGTHUNIT["metre",1],
            ID["EPSG",8807]]],
    CS[Cartesian,2],
        AXIS["(E)",east,
            ORDER[1],
            LENGTHUNIT["metre",1,
                ID["EPSG",9001]]],
        AXIS["(N)",north,
            ORDER[2],
            LENGTHUNIT["metre",1,
                ID["EPSG",9001]]]]
```
and dst.wkt, the same with PRIMEM["Reference Meridian",0]
```
PROJCRS["B",
    BASEGEOGCRS["Mercury",
        DATUM["Mercury",
            ELLIPSOID["Mercury",2439700,0,
                LENGTHUNIT["metre",1,
                    ID["EPSG",9001]]]],
        PRIMEM["Reference Meridian",0,
            ANGLEUNIT["degree",0.0174532925199433,
                ID["EPSG",9122]]]],
    CONVERSION["unknown",
        METHOD["Equidistant Cylindrical (Spherical)",
            ID["EPSG",1029]],
        PARAMETER["Latitude of 1st standard parallel",0,
            ANGLEUNIT["degree",0.0174532925199433],
            ID["EPSG",8823]],
        PARAMETER["Longitude of natural origin",180,
            ANGLEUNIT["degree",0.0174532925199433],
            ID["EPSG",8802]],
        PARAMETER["False easting",0,
            LENGTHUNIT["metre",1],
            ID["EPSG",8806]],
        PARAMETER["False northing",0,
            LENGTHUNIT["metre",1],
            ID["EPSG",8807]]],
    CS[Cartesian,2],
        AXIS["(E)",east,
            ORDER[1],
            LENGTHUNIT["metre",1,
                ID["EPSG",9001]]],
        AXIS["(N)",north,
            ORDER[2],
            LENGTHUNIT["metre",1,
                ID["EPSG",9001]]]]
```
then
```
$ projinfo -s @src.wkt -t @dst.wkt
unknown id, Inverse of unknown + Transformation from Mercury to Mercury + unknown, 0 m, World

PROJ string:
+proj=pipeline +step +inv +proj=eqc +lat_ts=0 +lat_0=0 +lon_0=180 +x_0=0 +y_0=0 +R=2439700 +pm=20 +step +proj=eqc +lat_ts=0 +lat_0=0 +lon_0=180 +x_0=0 +y_0=0 +R=2439700
```

**Comment by rouault** [comment]
Definition from ISO19111:2019
```
prime meridian
meridian from which the longitudes of other meridians are quantified
```
and later
```
If the subtype of Datum is geodetic reference frame, the description of the origin from which longitude values are specified – the prime meridian – is mandatory. 
```
EPSG guidance 7-2 also mentions
```
Longitudes are most commonly expressed relative to the Prime Meridian of Greenwich but some countries, particularly in former times, have preferred to relate their longitudes to a prime meridian through their national astronomic observatory, usually sited in or near their capital city, e.g. Paris for France, Bogota for Colombia.
```

> For a map projection to work, are Mercury PRIMEM["Reference meridian", 0.0] and PRIMEM["Hun Kal", 20] equivalent?

It depends on what you do exactly. If you transform between the base geographic CRS of this projected CRS and the projected CRS, since they use both the same prime meridian, you will get the same results whatever the PRIMEM value (since it cancels itself). But if you transform from a CRS with PRIMEM["Hun Kal", 20] to a CRS with  PRIMEM["Reference meridian", 0.0], then you'll get a 20deg shift.
If you have a geographic CRS with PRIMEM["Hun Kal", 20], it means that if you pass a coordinate expressed in that CRS whose longitude value is 0, it is 0 relative to Hun Kal = 20.
Honestly, if you have the choice, using a Reference_Meridian = 0 looks to me the simplest solution since it will remove a number of potential confusion. Non 0 prime meridian are mostly an artifact of historical Earth geodesy. How this 0 is defined relatively to observations might indeed be added as a REMARK.

**Comment by desruisseaux** [comment]
My problem with "Reference meridian" is that the name is quite vague about what this meridian is (what we call "reference" may vary with peoples and times), while Hun Kal is much more explicit (on Earth, we have "Greenwich" which is quite clear).

However I realize that `PRIMEM["Hun Kal", 20]` may not be appropriate since it would mean that `GeographicCRS` using that prime-meridian would have longitude 0 on the Hun Kal crater. If I understood correctly, this is not the intent (the indent is that the longitude of Hun Kal crater stay 20°), is that right? If yes, then I think that my recommendation to use `PRIMEM["Hun Kal", 20]` was wrong. I would just recommend to use a name more explicit than "Reference Meridian" for the meridian at 0°. Could be "Mercury reference meridian", but I think it would be better if the name had some reference to the adoption as a reference (e.g. international organization + year when it was decided to base the definition on Hun Kal).

**Comment by thareUSGS** [comment]
Okay I think we are converging on:

1. The use Hun Kal is seemingly **not** a valid substitute for the Prime meridian in the `GeographicCRS`. 
2. Perhaps adding a REMARK for Hun Kal would be good. (e.g. I don't know maybe `["REMARK", "Hun Kal, IAU2015 longitude reference point at 20W"]`.   Seems a bit long? Do we need it?
3. I agree that "Reference Meridian" is very generic. Now I have been told this string can be anything. But we have started to add `AUTHORITY["IAU2015","19900"]` to the GeographicCRS which is intended to cover the whole definition (as pointing to the IAU publication in 2015). Thus perhaps keeping it simply "Reference Meridian" is enough?

I came to this conclusion with help from Brent Archinal (from the IAU). Like us, he was also initially somewhat confused, but more from not understanding WKT.

Here are some of his points:

1. A zero meridian, reference meridian, and prime meridian are all the same thing.
2. There may be a feature that defines longitude, but obviously it doesn't need to be at zero.  **That doesn't mean the longitude defining feature is the reference meridian.**
* He also points out:    Airy-0 is **now** not at zero for Mars either. Viking 1 provides the current longitude definition.
3. ...the reason to have a defining feature is to check whether that transformation needs to be updated after newer / more accurate data is obtained.  It  doesn't matter if you're map projecting data that already is in body fixed coordinates. To put it another way, if the feature ends up with the wrong coordinates on your map, the error is in the inertial to body fixed transformation, not the map projection.
4. On Earth, the real reason the Reference Meridian parameter being non-zero is to allow reference to historical prime meridians for the Earth which have an offset from the current zero degree meridian. E.g. before the [International Meridian Conference of 1884](https://en.wikipedia.org/wiki/International_Meridian_Conference) pretty much each country had their own meridian_...examples listed..._
5. And finally - this is a good one for Earth-users :-) -- The current internationally accepted name of the Earth's prime meridian is the IERS Reference Meridian (IRM) (https://en.wikipedia.org/wiki/IERS_Reference_Meridian).  It is **not** the same as the Greenwich meridian, which is about 100 meters west of the IRM.

**Comment by rouault** [comment]
As an alternative to REMARK, I was wondering if you might not use the ANCHOR[] sub-node of DATUM[] allowed in WKT2 that models the datum anchor definition of ISO191111 : "
description, possibly including coordinates of an identified point or points, of the relationship used to anchor a coordinate system to the Earth or alternate object. Note: For modern geodetic reference frames the anchor may be a set of station coordinates; if the reference frame is dynamic it will 	also include coordinate velocities. For a traditional geodetic datum, the anchor may be a point known as the fundamental point, which is traditionally the point where the relationship between geoid and ellipsoid is defined, together with a direction from that point. "
Look for "Tananarive 1925" in http://docs.opengeospatial.org/is/12-063r5/12-063r5.html

**Comment by thareUSGS** [comment]
hmm - using ANCHOR to define Hun Kal sounds **very tempting.** We will need to dig into that a little more to see what are the implications for the map projection. Also using a coordinate velocity might be a solution for the planet's spin rate (but that sounds dangerous as it is outside the intended use).

**Comment by J-Christophe** [comment]
Okay, it is much clearer for me now.

Concerning the point 2 : we do not really need it (that's a remark now :-) ). However this particular feature on the map is an information and it can help to check the accuracy of the longitude on the map. That's not bad. In another hand if we keep the remark, we have to be careful because the West/East depends on ocentric/ographic coordinate system. In ocentric, the longitude is always counted positively to East where as the longitude is counted positively to East/West according to the rotation sens of the planet for ographic. So according to the coordinate system, the remark should change.

Concerning the point 3 : The reference meridian is fixed for each planet and it will not evolve in the time. From [IAU report](https://link.springer.com/epdf/10.1007/s10569-017-9805-5?author_access_token=pcbs1gV_DYhP9yhmAYezFPe4RwlQNchNByi7wbcMAY6HRRN_JGkCH_CXrmP6w-LLXlanSQWWDo6B6Yb0vVQo5UJybROqEAn3dWfKapiaqajv2hJnIdrYZz0dRSCoaVszFbGvpOxn-3Iboajxlxffmw%3D%3D) : 

> For planets or satellites with no accurately observable ﬁxed surface features, the expression for W deﬁnes the **prime meridian and is not subject to correction** for this reason

> Where the cartographic position of the prime meridian is deﬁned by an observable feature, the expression for W is chosen so that the ephemeris position follows the motion of the feature as closely as possible. **When higher accuracy mapping** is done or a new value for W is derived, **the longitude of the deﬁning feature must be maintained**.

> once an observable feature is chosen to deﬁne the longitude system, that system should not change except under extraordinary circumstances. For example,a few solar system bodies such as Io and various comet nuclei are known to be geologically active at the present time, so it is conceivable that surface processes could render a small reference feature unidentiﬁable and necessitate the choice of a new (but consistent) reference.

So for me keeping "Reference meridian" with AUTHORITY["IAU2015","19900"] is largely enough and simple.

**Comment by thareUSGS** [comment]
@J-Christophe can you take a stab at updating the code for WKT1 so we can close this monster? I think adding WKT2 support (and maybe testing the addition of ANCHOR) can be a secondary task.

**Comment by J-Christophe** [comment]
I just started doing it 10 mn ago :-)

Le mer. 6 févr. 2019 à 22:36, Trent Hare <notifications@github.com> a
écrit :

**Comment by SamuelAndresPascal** [comment]
I'm quite new to this kind of subjects, but I thought about Hun Kal case those last days.

I would like to contrast two cases without consideration for any syntactic issue. The first case is the piece of information **('Paris', 2.5969213)** and the second one the piece of information **('Hun Kal', 20.)**.

The semantics of the numerical value seems to be very different in both cases.

In the first case, the parameter is a _longitude offset of the currently defined meridian, from the meridian referenced by all the prime meridians of the body_ (we could almost call such a meridian "Greenwich meridian" in a generic way, even for other bodies than Earth). The purpose of this offset is to allow prime meridian changes from one to another. It is expressed in a kind of meridian "angular reference system" associated to the Greenwich meridian, and not in the meridian "angular reference system" associated to the meridian currently described (Paris).

(Please forgive both my English and my own words with which I try to describe every concepts.)

In the second case, the parameter is _not_ a longitude offset from another known and more fundamental "prime meridian". Its purpose is _not_ to allow prime meridian changes but _to locate the currently defined prime meridian_ in its own "angular reference system". It is why this parameter is expressed in the "angular reference system" of the very currently described (Hun Kal) meridian. I think we can give different interpretations about this. For example :

**a)** As a first interpretation, we could view there is none obligation for the so-called "Hun Kal" meridian to pass through the Hun Kal crater. It would be just a meridian positioned _"relatively to"_ the Hun Kal crater, at -20° from it. In such an interpretation the "Hun Kal prime meridian" is defined by the Hun Kal crater, but it does not pass through it. Its longitude is 0° and the Hun Kal crater one is 20°.

**b)** If we absolutely want the so-called "Hun Kal" meridian to pass through the Hun Kal crater, this means the 0° meridian in the Hun Kal meridian "angular reference system" is different from the prime meridian itself whose longitude in this system is 20°. In such an interpretation, the "Hun Kal prime meridian" is defined by the Hun Kal crater and it passes through it. It's longitude is 20° which means the prime meridian is not the origin of the longitudes.

Both interpretations seems interesting to me because both forces to redefine the concepts we are "familiar" with when addressing terrestrial information. But I think the common concepts we daily use may need to be generalized or revisited to address both more specific and more general cases in planetary information. In the case of "Hun Kal prime meridian" it seems the _"prime meridian"_ concept can be different from the _"reference meridian"_ (I mean, the meridian that passes through the physical reference) and from the _"longitude origin"_ (I mean, the meridian whose longitude is 0°). The Hun Kal case forces us to choose the prime meridian to be located either at the "longitude origin" (so not at the "reference meridian", case **a)**), or at the reference meridian (so not at the "longitude origin", case **b)**). It is a question of convention.


To summarize, I think interesting to think about at least two different parameters to correctly describe the "Hun Kal case" :

1) The longitude of the currently described prime meridian from the equivalent of the "Greenwich meridian" for the body.
2) The longitude of the currently described prime meridian in the meridian "angular reference system" which it defines itself (or the equivalent longitude at which the physical reference is located, according to the chosen interpretation **a)** or **b)**).

Such parameters would be, for the following different cases :
('Greenwich', 0., 0.)
('Paris', 2.5969213, 0.)
('Hun Kal', 0., 20.)

Here we are coherent with the fact that 'Hun Kal' prime meridian is not defined from a more fundamental prime meridian, but from the physical reference located at +20° from the longitude origin.

Since on Earth all the prime meridians seems to be both "reference meridians" and "longitude origins", the second parameter is implicitly 0. So, in WKT, the second parameter does not already exist and these notations would become in my opinion :
["Greenwich", 0.]
["Paris", 2.5969213]
["Hun Kal", 0.]


We could even imagine cases with 3 different locations for _longitude origin_, _prime meridian_ and _reference meridian_. In think such cases would require one more parameter.

May be I am wrong or may be I am telling about evidences ? But from my point of view, since the problem appears to be conceptual to me, I think it would be a semantic innovation to write ["Hun Kal", 20.] or ["Reference Meridian", 0.]. It implicitly (in the first case) or explicitly (in the second case) refers to a "Reference Meridian" that is quite artificial.

Best regards for all.

**Comment by thareUSGS** [comment]
fixed with #36
